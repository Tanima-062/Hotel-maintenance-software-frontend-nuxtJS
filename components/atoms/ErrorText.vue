<template>
  <transition name="fade">
    <p
      v-if="!!errText"
      :class="`text-alert ${NoWrap ? 'whitespace-no-wrap' : ''}`"
    >
      {{ errText }}
    </p>
  </transition>
</template>

<script lang="ts">
import { Component, Vue, Prop } from 'vue-property-decorator'

@Component
export default class extends Vue {
  @Prop({ default: '' })
  errText!: string

  @Prop({ default: false })
  NoWrap!: boolean
}
</script>
<style lang="sass" scoped>
.fade-enter-active, .fade-leave-active
  transition: opacity .2s

.fade-enter, .fade-leave-to /* .fade-leave-active below version 2.1.8 */
  opacity: 0
</style>
