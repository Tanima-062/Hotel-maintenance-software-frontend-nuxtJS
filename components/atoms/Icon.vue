<template>
  <div :class="styles">
    <svg
      class="w-7 h-7 fill-current"
      :class="`${color}`"
      :style="`width: ${iconWidth}px; height: ${iconHeight}px`"
      :viewBox="`0 0 24 24`"
    >
      <path :d="icon" />
    </svg>
  </div>
</template>
<script lang="ts">
import { Vue, Component, Prop } from 'nuxt-property-decorator'

@Component
export default class extends Vue {
  @Prop({ default: '' })
  icon!: string

  @Prop({ default: 'text-white' })
  color!: string

  @Prop({ default: '' })
  bgColor!: string

  @Prop({ default: false })
  round!: boolean

  @Prop({ default: 24 })
  iconWidth!: number

  @Prop({ default: 24 })
  iconHeight!: number

  @Prop({ default: '' })
  extraClass!: string

  get styles(): string {
    const classList = []
    if (this.bgColor) {
      classList.push(`${this.bgColor}`)
    }
    if (this.round) {
      classList.push('rounded-full')
    }
    if (this.extraClass) {
      classList.push(this.extraClass)
    }
    return classList.join(' ')
  }
}
</script>
