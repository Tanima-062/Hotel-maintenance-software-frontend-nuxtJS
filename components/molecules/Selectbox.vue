<template>
  <validation-provider
    v-slot="{ errors }"
    tag="div"
    :rules="rules"
    :name="name"
    :class="width"
  >
    <div class="relative inline-block text-left" :class="width">
      <transition name="selectbox">
        <div
          v-show="clicked"
          class="origin-bottom-right absolute right-0 mt-2 w-full rounded-md shadow-lg bg-white z-10"
          :class="`${boxStyle}`"
        >
          <div class="py-1" role="menu" v-if="!isDisabled">
            <a
              v-for="(item, key) in items"
              :key="key"
              class="block px-4 py-2 text-sm text-black hover:bg-menuColor"
              :class="width"
              role="menuitem"
              @mousedown="
                isSelectFirstItem
                  ? selectedItemOption(item.value, key)
                  : selectItem(item.value) && selectItemIndex(key)
              "
            >
              {{ item.name }}
            </a>
          </div>
        </div>
      </transition>
      <div>
        <button
          id="options-menu"
          type="button"
          class="flex items-center justify-between w-full rounded-md text-selectColor-icon"
          :class="`${width} ${padding} ${border}`"
          :error="!!errors[0]"
          aria-haspopup="true"
          aria-expanded="true"
          @click="clicked = !clicked"
          @blur="clicked = false"
        >
          <p
            class="text-left"
            :class="activeItem.value === 0 ? 'text-black-light' : ''"
          >
            {{ activeItem.name }}
          </p>
          <Icon
            v-if="!isDisabled"
            class="flex flex-wrap -mr-1 ml-2"
            :icon="downIcon"
            color="text-black"
          />
        </button>
      </div>
    </div>
    <ErrorText :err-text="errors[0]" />
  </validation-provider>
</template>
<script lang="ts">
import { Vue, Component, Prop, Emit } from 'nuxt-property-decorator'
import { mdiMenuDown } from '@mdi/js'
import { Global } from '~/types'

@Component({
  components: {
    Icon: () => import('~/components/atoms/Icon.vue'),
    ErrorText: () => import('~/components/atoms/ErrorText.vue'),
  },
})
export default class extends Vue {
  @Prop({ default: false })
  isDisabled!: boolean

  @Prop({ default: [], required: true })
  items!: Global.SelectboxItem[]

  @Prop({ default: {}, required: true })
  activeItem!: Global.SelectboxItem

  @Prop({ default: 'top' })
  boxPosition!: string

  @Prop({ default: '' })
  width!: string

  @Prop({ default: 'p-1' })
  padding!: string

  @Prop({ default: '' })
  maxHeight!: string

  @Prop({ default: '' })
  name!: string

  @Prop({ default: '' })
  rules!: string

  @Prop({ default: false })
  error!: boolean

  @Prop({ default: '' })
  boxHeight!: string

  @Prop({ default: false })
  isSelectFirstItem!: boolean

  private clicked: boolean = false
  private downIcon: string = mdiMenuDown

  get boxStyle(): string {
    const styleArray = []
    if (this.boxPosition === 'top') {
      styleArray.push('top-0')
    }
    if (this.boxPosition === 'bottom') {
      styleArray.push('bottom-0')
    }
    if (this.maxHeight) {
      styleArray.push(this.maxHeight)
      styleArray.push('overflow-auto')
    }
    styleArray.push(this.boxHeight)
    styleArray.push('overflow-auto')

    return styleArray.join(' ')
  }

  get border(): string {
    const selectBox = []
    selectBox.push(
      this.error
        ? 'border border-alert bg-error'
        : 'border border-solid border-bdColor bg-white'
    )
    return selectBox.join(' ')
  }

  private selectedItemOption(value: number | string, index: number) {
    this.clicked = false
    this.selectItem(value)
    this.selectItemIndex(index)
  }

  @Emit()
  private selectItem(value: number | string): number | string {
    this.clicked = false
    return value
  }

  @Emit()
  private selectItemIndex(index: number): number | string {
    this.clicked = false
    return index
  }
}
</script>
<style lang="sass" scoped>
.selectbox-enter-active, .selectbox-leave-active
  transition: opacity .1s

.selectbox-enter, .selectbox-leave-to
  opacity: 0
</style>
