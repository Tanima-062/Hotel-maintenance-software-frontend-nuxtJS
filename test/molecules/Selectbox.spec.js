import { mount } from '@vue/test-utils'
import Selectbox from '@/components/molecules/Selectbox.vue'
import Icon from '@/components/atoms/Icon.vue'
import ErrorText from '@/components/atoms/ErrorText.vue'

describe('Molecules Selectbox Check', () => {
  test('Selectbox Test', () => {
    const props = {
      items: [
        {
          name: 'test 1',
          value: 'test 1',
        },
        {
          name: 'test 2',
          value: 'test 2',
        },
      ],
      activeItem: {
        name: 'test 1',
        value: 'test 1',
      },
    }
    const wrapper = mount(Selectbox, {
      propsData: props,
      stubs: {
        Icon,
        ErrorText,
      },
    })
    // check if it is a vue element
    expect(wrapper.vm).toBeTruthy()

    // check Snapshot
    expect(wrapper.element).toMatchSnapshot()
  })
})
