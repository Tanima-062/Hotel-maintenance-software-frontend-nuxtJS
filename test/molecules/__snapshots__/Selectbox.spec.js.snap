// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Molecules Selectbox Check Selectbox Test 1`] = `
<div
  class=""
>
  <div
    class="relative inline-block text-left"
  >
    <transition-stub
      name="selectbox"
    >
      <div
        class="origin-bottom-right absolute right-0 mt-2 w-full rounded-md shadow-lg bg-white z-10 top-0  overflow-auto"
        style="display: none;"
      >
        <div
          class="py-1"
          role="menu"
        >
          <a
            class="block px-4 py-2 text-sm text-black hover:bg-menuColor cursor-pointer"
            role="menuitem"
          >
            
            test 1
          
          </a>
          <a
            class="block px-4 py-2 text-sm text-black hover:bg-menuColor cursor-pointer"
            role="menuitem"
          >
            
            test 2
          
          </a>
        </div>
      </div>
    </transition-stub>
     
    <div>
      <button
        aria-expanded="true"
        aria-haspopup="true"
        class="flex items-center justify-between w-full rounded-md text-selectColor-icon  p-1 border border-solid border-bdColor bg-white"
        id="options-menu"
        type="button"
      >
        <p
          class="text-left"
        >
          
          test 1
        
        </p>
         
        <div
          class="flex flex-wrap -mr-1 ml-2"
        >
          <svg
            class="w-7 h-7 fill-current text-black"
            style="width: 24px; height: 24px;"
            viewBox="0 0 24 24"
          >
            <path
              d="M7,10L12,15L17,10H7Z"
            />
          </svg>
        </div>
      </button>
    </div>
  </div>
   
  <!---->
</div>
`;
